### PR TITLE
Rpi eeprom upstream

### PR DESCRIPTION
**Pull request by macpijan**
Fixes: https://github.com/agherzan/meta-raspberrypi/issues/515

**- What I did**
- added rpi-eeprom recipe to allow EEPROM firmware update on the RPI4 devices
- added userlandtools recipe, which provides part of the userland tools required for EEPROM update

**- How I did it**
- inspired by the recipes from Balena
- more detailed description in the commit messages
- tested only on dunfell so far, that is why this PR targets dunfell

### CODE REVIEW

**Comment by macpijan** [comment]
I wanted to package the rpi-eeprom-config separately (as it pulls in Python), but it seems that the rpi-epprom-update also calls it, so the whole rpi-eeprom would need to RDEPENDS on Python most likely https://github.com/raspberrypi/rpi-eeprom/blob/master/rpi-eeprom-update#L139

File: recipes-bsp/rpi-eeprom/files/0001-rpi-eeprom-config-use-usr-bin-python.patch [hunk 1]
```diff
@@ -0,0 +1,22 @@
+From 4b806a7be5c28e5a76a26f899b6be0500b39d718 Mon Sep 17 00:00:00 2001
+From: leckijakub <jakub.lecki@3mdeb.com>
+Date: Tue, 8 Sep 2020 12:35:42 +0200
+Subject: [PATCH] rpi-eeprom-config: use #!/usr/bin/python
+
+Signed-off-by: leckijakub <jakub.lecki@3mdeb.com>
```
Comment (agherzan) [inline recipes-bsp/rpi-eeprom/files/0001-rpi-eeprom-config-use-usr-bin-python.patch:6]: Upstream status missing.

File: recipes-bsp/rpi-eeprom/files/0001-rpi-eeprom-config-use-usr-bin-python.patch [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ rpi-eeprom-config | 2 +-
+ 1 file changed, 1 insertion(+), 1 deletion(-)
+
+diff --git a/rpi-eeprom-config b/rpi-eeprom-config
+index af0d63b11b75..a6b5c4d29b12 100755
+--- a/rpi-eeprom-config
++++ b/rpi-eeprom-config
+@@ -1,4 +1,4 @@
+-#!/usr/bin/env python
++#!/usr/bin/env python3
```
Comment (agherzan) [inline recipes-bsp/rpi-eeprom/files/0001-rpi-eeprom-config-use-usr-bin-python.patch:17]: This doesn't seem to be what you are advertising in your commit log. Also, there is no description of the problem this patch is fixing (even though I know what it is).

File: recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb [hunk 3]
```diff
@@ -0,0 +1,71 @@
+FILESEXTRAPATHS_append := ":${THISDIR}/files"
```
Comment (agherzan) [inline recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb:1]: This is not needed.

File: recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
```
Comment (agherzan) [inline recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb:2]: Summary, please.

File: recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+DESCRIPTION = "This repository contains the rpi4 bootloader and scripts \
+for updating it in the spi eeprom"
+LICENSE = "BSD-3-Clause"
+LIC_FILES_CHKSUM = "file://LICENSE;md5=7dcd1a1eb18ae569857c21cae81347cb"
+
+# EEPROM configuration default values for this version, as specified at
+# https://www.raspberrypi.org/documentation/hardware/raspberrypi/booteeprom.md
```
Comment (agherzan) [inline recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb:9]: I'm not sure what is this referring to. The next line is not a configuration - it's the source's repository.

File: recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+SRC_URI = " \
+    git://github.com/${SRCFORK}/rpi-eeprom.git;protocol=git;branch=${SRCBRANCH} \
+    file://0001-rpi-eeprom-config-use-usr-bin-python.patch \
+"
+
+COMPATIBLE_MACHINE = "raspberrypi4-64"
```
Comment (agherzan) [inline recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb:15]: This, by convention, is mentioned at the end of the recipe.

File: recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+DEPENDS = "python3-native"
```
Comment (agherzan) [inline recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb:17]: Inheriting `python3native` would be better. But is this dependency actually needed for the build?

EDIT: actually you do it later. So this is not needed anymore.

File: recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+RDEPENDS_${PN} = "coreutils userlandtools"
+RDEPENDS_${PN}-config = "python3"
+
+SRCBRANCH = "master"
```
Comment (agherzan) [inline recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb:22]: Move this above SRC_URI to keep the same context.

File: recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+SRCFORK = "raspberrypi"
+SRCREV = "09f77ad9fa8655a28de754640e82ca67aabe6c33"
+
+# Use the date of the above commit as the package version. Update this when
+# SRCREV is changed.
+PV = "20200903"
+
+S = "${WORKDIR}/git"
```
Comment (agherzan) [inline recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb:30]: Move this after SRC_URI.

File: recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    install -D ${S}/rpi-eeprom-update-default ${D}${sysconfdir}/default/rpi-eeprom-update
+}
+
+PACKAGES =+ "${PN}-config"
+FILES_${PN}-config = "${bindir}/rpi-eeprom-config"
+
+FILES_${PN} += " ${sbindir}/vl805"
+FILES_${PN} += "${base_libdir}/firmware/raspberrypi/bootloader/*"
+FILES_${PN} += "${sysconfdir}/*"
+
```
Comment (agherzan) [inline recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb:62]: Extra new line.

File: recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+INHIBIT_PACKAGE_STRIP = "1"
+INHIBIT_PACKAGE_DEBUG_SPLIT = "1"
+
+# vl805 tool sources are not available (yet),
```
Comment (agherzan) [inline recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb:67]: The line-length looks unecesarry short.

File: recipes-graphics/userland/userland.inc [hunk 12]
```diff
@@ -0,0 +1,38 @@
+DESCRIPTION = "This repository contains the source code for the ARM side \
```
Comment (agherzan) [inline recipes-graphics/userland/userland.inc:1]: Can't we just have a separate package inside the userland recipe for the EEPROM tools? Wouldn't userland compile?

File: recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb [hunk 13]
```diff
@@ -15,7 +15,8 @@ SRC_URI = " \
 COMPATIBLE_MACHINE = "raspberrypi4-64"
 
 DEPENDS = "python3-native"
-RDEPENDS_${PN} = "coreutils userlandtools"
+RDEPENDS_${PN} = "coreutils"
```
Comment (agherzan) [inline recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb:18]: Merge this in the initial commit.

**Review by agherzan — CHANGES_REQUESTED** [review]

**Comment by xlla** [comment]
I have pull&merge this PR into my workspace, encounter those errors while build rpi-eeprom.
<pre>
ERROR: userlandtools-20201027-r0 do_package: QA Issue: userlandtools: Files/directories were installed but not shipped in any package:
  /usr/lib/plugins/writer_avi.so
  /usr/lib/plugins/reader_flv.so
  /usr/lib/plugins/writer_dummy.so
  /usr/lib/plugins/reader_mpga.so
  /usr/lib/plugins/writer_asf.so
  /usr/lib/plugins/reader_simple.so
  /usr/lib/plugins/reader_rcv.so
  /usr/lib/plugins/reader_rtp.so
  /usr/lib/plugins/reader_raw_video.so
  /usr/lib/plugins/reader_asf.so
  /usr/lib/plugins/reader_binary.so
  /usr/lib/plugins/writer_binary.so
  /usr/lib/plugins/reader_rv9.so
  /usr/lib/plugins/writer_mp4.so
  /usr/lib/plugins/reader_ps.so
  /usr/lib/plugins/reader_qsynth.so
  /usr/lib/plugins/reader_mp4.so
  /usr/lib/plugins/reader_metadata_id3.so
  /usr/lib/plugins/reader_rtsp.so
  /usr/lib/plugins/writer_simple.so
  /usr/lib/plugins/reader_wav.so
  /usr/lib/plugins/writer_raw_video.so
  /usr/lib/plugins/reader_mkv.so
  /usr/lib/plugins/reader_avi.so
Please set FILES such that these items are packaged. Alternatively if they are unneeded, avoid installing them or delete them within do_install.
userlandtools: 24 installed and not shipped files. [installed-vs-shipped]
ERROR: userlandtools-20201027-r0 do_package: Fatal QA errors found, failing task.
</pre>

compare to userland_git, "${libdir}/plugins" should be include or delete.

**Comment by xlla** [comment]
I do not know why rpi-eeprom depends userlandtools instead of userland.
Both userlandtools and userland will generate same *.so and install to same place, but those libs was build with different patched source and  different configs.

File: recipes-graphics/userland/userland.inc (hunk 12 above)
Comment (kraj) [reply recipes-graphics/userland/userland.inc:1]: we don't include userland when choosing vc4graphics, so a separate package for these tools is a fine approach.

File: recipes-graphics/userland/userland.inc [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+
+COMPATIBLE_MACHINE = "^rpi$"
+
+SRCBRANCH = "master"
+SRCFORK = "raspberrypi"
+SRCREV = "188d3bfe4a0ac36b119a2cee35a6be8d0c68e09e"
+
+# Use the date of the above commit as the package version. Update this when
+# SRCREV is changed.
+PV = "20200624"
```
Comment (kraj) [inline recipes-graphics/userland/userland.inc:16]: Please abstract common bits between userland recipe and this recipe into an inc file e.g. all above this line

File: recipes-graphics/userland/userlandtools_git.bb [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+        rm -rf ${D}${bindir}/vchiq_test
+        rm -rf ${D}${bindir}/dtmerge
+        rm -rf ${D}${prefix}/include
+        rm -rf ${D}${prefix}/src
+        rm -rf ${D}${libdir}/pkgconfig
+        rm -rf ${D}${libdir}/*debug*
+        rm -rf ${D}${libdir}/*host*
+        rm -rf ${D}${libdir}/*.a
+        rm -rf ${D}${bindir}/*post
+        rm -rf ${D}${bindir}/*pre
```
Comment (kraj) [inline recipes-graphics/userland/userlandtools_git.bb:20]: so many rm cmds in series, can we move the ones we need out and delete everything and recreate install of whats really needed.

**Review by kraj — CHANGES_REQUESTED** [review]

File: recipes-graphics/userland/userland.inc (hunk 14 above)
Comment (macpijan) [reply recipes-graphics/userland/userland.inc:16]: It already **is** a .inc file you are referring to. Or maybe I'm missing what you mean.

File: recipes-graphics/userland/userlandtools_git.bb (hunk 15 above)
Comment (macpijan) [reply recipes-graphics/userland/userlandtools_git.bb]: @kraj Updated according to the suggestion. I hope this is what you meant.

File: recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb (hunk 3 above)
Comment (macpijan) [reply recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb]: @agherzan I hope all your suggestions are fixed now.

File: recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+
+INHIBIT_PACKAGE_STRIP = "1"
+INHIBIT_PACKAGE_DEBUG_SPLIT = "1"
+
+# vl805 tool sources are not available (yet), as it comes as a precompiled
+# binary only. It has ARM architecture whereas target machine is Aarch64. We
+# need to disable arch check for it otherwise it cannot packed.
+QAPATHTEST[arch] = ""
+
+COMPATIBLE_MACHINE = "raspberrypi4-64"
```
Comment (agherzan) [inline recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb:62]: Only compatible for 64bit?

File: recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb [hunk 17]
```diff
@@ -0,0 +1,62 @@
+SUMMARY = "Installation scripts and binaries for the Raspberry Pi 4 EEPROM"
+DESCRIPTION = "This repository contains the rpi4 bootloader and scripts \
+for updating it in the spi eeprom"
+LICENSE = "BSD-3-Clause"
+LIC_FILES_CHKSUM = "file://LICENSE;md5=7dcd1a1eb18ae569857c21cae81347cb"
+
+SRCBRANCH = "master"
+SRCFORK = "raspberrypi"
```
Comment (agherzan) [inline recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb:8]: I think both of these are superfluous. The branch is master by default.

File: recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+    git://github.com/${SRCFORK}/rpi-eeprom.git;protocol=git;branch=${SRCBRANCH} \
+    file://0001-rpi-eeprom-config-convert-sheabang-from-python-to-py.patch \
+"
+
+S = "${WORKDIR}/git"
+
+RDEPENDS_${PN} = "coreutils python3"
+RDEPENDS_${PN} += "${@bb.utils.contains("MACHINE_FEATURES", "vc4graphics", "userlandtools", "userland", d)}"
+
+SRCREV = "09f77ad9fa8655a28de754640e82ca67aabe6c33"
```
Comment (agherzan) [inline recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb:20]: NIT: Move this in the same section with `S` and `SRC_URI`.

File: recipes-graphics/userland/userland.inc (hunk 12 above)
Comment (agherzan) [reply recipes-graphics/userland/userland.inc:1]: Good point.

File: recipes-graphics/userland/userland.inc (hunk 14 above)
Comment (agherzan) [reply recipes-graphics/userland/userland.inc:16]: I think @kraj missed that this is a `inc` and not a separate `bb`.

File: recipes-graphics/userland/userlandtools_git.bb [hunk 19]
```diff
@@ -0,0 +1,33 @@
+require userland.inc
+
+RCONFLICTS_${PN} = "userland"
+
+# Keep only those libs & bins that are actually used during EEPROM update
+do_install_append () {
```
Comment (agherzan) [inline recipes-graphics/userland/userlandtools_git.bb:6]: This looks a bit hacky - can't we just overwrite the install task manually pulling in only the needed blobs?

**Review by agherzan — CHANGES_REQUESTED** [review]
Some small comments here and there. I think we should be ready to merge afterwards.

File: recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb (hunk 3 above)
Comment (agherzan) [reply recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb]: Thanks. Added some more comments. Also, a rebase is needed (and a test).

**Comment by om26er** [comment]
We have a need to get this stuff working. I am going to work on something quick that we could use internally and then at a later stage try to add those recipes to `meta-raspberrypi` as well.

**Comment by hommeabeil** [comment]
I need it too @agherzan what are the best way to take over this PR ? Do I re-open one or I can contribute directly in this one ?

**Comment by macpijan** [comment]
I can help to finish it up to some extent, but probably not enough time to test right now. I fixed some simple review comments. This one is left if you wish to continue with that: https://github.com/agherzan/meta-raspberrypi/pull/741#discussion_r538408275

I think the way to do that would be to build existing code, see what goes into the package and pick only the necessary bits in the custom do_install as @agherzan have suggested.

I think we can continue using this PR. You could issue PR to my branch https://github.com/3mdeb/meta-raspberrypi/tree/rpi-eeprom-upstream and I will merge it.

**Comment by agherzan** [comment]
There was no activity on this for a while. I might look into this soon. In the meanwhile, I'll close this MR.

**Pull request closed without merging**